File: 886-score-of-parentheses/score-of-parentheses.cpp
class Solution {
public:
    int scoreOfParentheses(string s) {
        stack<int> st;

        // Score of the outermost level
        st.push(0);

        for (char c : s) {

            if (c == '(') {
                st.push(0);
            } 
            else {
                // Score inside the current pair
                int curr = st.top();
                st.pop();

                // A  = () -> 1
                // (A) -> 2 * score(A)
                int value = (curr == 0) ? 1 : 2 * curr;

                // Add the current score to the parent level
                int parent = st.top();
                st.pop();

                st.push(parent + value);
            }
        }

        return st.top();
    }
};

File: 886-score-of-parentheses/score-of-parentheses.java
class Solution {
    public int scoreOfParentheses(String s)
    {
        Stack<Integer> st = new Stack<>();
        st.push(0);
        for(int i=0;i<s.length();i++)
        {
            if(s.charAt(i) == '(')
                st.push(0);
            else
            {
                int temp = st.peek();
                st.pop();
                
                int count = temp==0? 1 : 2*temp;
                int parent = st.peek();
                st.pop();

                st.push(count+parent);
            }
        }

        return st.peek();    
    }
}